lib/index.js

```diff
@@ -19,23 +31,47 @@ module.exports = (referenceDir, config, { watch, fingerprint, sourcemaps, compac
 		fingerprint,
 		exitOnError: !watch
 	});
-	let watcher = watch && makeWatcher(config.watchDirs, referenceDir,
-			assetManager.resolvePath);
 	let browsers = browserslist.findConfig(referenceDir) || {};
 
 	let plugins = Object.assign({}, DEFAULTS.plugins, config.plugins);
+	let buckets = {
+		static: [],
+		scripts: [],
+		styles: [],
+		markup: []
+	};
 	Object.keys(plugins).forEach(type => {
 		let pluginConfig = config[type];
 		if(!pluginConfig) {
 			return;
 		}
 
-		let plugin = plugins[type];
+		let bucket = plugins[type].bucket;
+		let plugin = plugins[type].package;
 		if(!plugin.call) {
 			plugin = load(plugin);
 		}
-		plugin(pluginConfig, assetManager, { watcher, browsers, sourcemaps, compact });
+		let initializedPlugin = plugin(pluginConfig, assetManager,
+				{ browsers, sourcemaps, compact });
+		buckets[bucket].push(initializedPlugin);
 	});
+
+	let buildStep1 = buckets.static;
+	let buildStep2 = buckets.scripts.concat(buckets.styles);
+	let buildStep3 = buckets.markup;
+
+	let buildAll = files => {
+		return Promise.all(buildStep1.map(plugin => plugin(files))).
+			then(() => Promise.all(buildStep2.map(plugin => plugin(files)))).
+			then(() => Promise.all(buildStep3.map(plugin => plugin(files))));
+	};
+	buildAll();
+
+	if(watch) {
+		let watcher = makeWatcher(config.watchDirs, referenceDir,
+				assetManager.resolvePath);
+		watcher.on("edit", buildAll);
```

Comment:
Minor nit: Seems like that `watcher` variable is unnecessary; readability-wise, chaining seems just as good:

```javascript
makeWatcher(config.watchDirs, referenceDir, assetManager.resolvePath).
    on("edit", buildAll);
```